metron-analytics/metron-profiler-common/src/main/java/org/apache/metron/profiler/hbase/DecodableRowKeyBuilder.java

```diff
@@ -0,0 +1,402 @@
+
+    return buffer.array();
+  }
+
+  /**
+   * Decodes a row key to build a ProfileMeasurement containing all of the
+   * relevant fields that exist within the row key.
+   *
+   * @param rowKey The row key to decode.
+   * @return A ProfileMeasurement.
+   */
+  @Override
+  public ProfileMeasurement decode(byte[] rowKey) {
+    ByteBuffer buffer = ByteBuffer
+            .wrap(rowKey)
+            .order(byteOrder);
+
+    try {
+      // validate the magic number
+      short magicNumber = buffer.getShort();
+      if(magicNumber != MAGIC_NUMBER) {
+        throw new IllegalArgumentException(String.format("Invalid magic number; expected '%s', got '%s'", MAGIC_NUMBER, magicNumber));
+      }
+
+      // validate the row key version
+      byte version = buffer.get();
+      if(version != VERSION) {
+        throw new IllegalArgumentException(String.format("Invalid version; expected '%s', got '%s'", VERSION, version));
+      }
+
+      // validate the salt length
+      int saltLength = buffer.getInt();
+      if (saltLength <= 0 || saltLength > MAX_FIELD_LENGTH) {
+        throw new IllegalArgumentException(String.format("Invalid salt length; max allowed '%d', got '%d'", MAX_FIELD_LENGTH, saltLength));
+      }
+
+      // decode the salt
+      byte[] salt = new byte[saltLength];
+      buffer.get(salt);
+
+      // validate the profile length
+      int profileLength = buffer.getInt();
+      if (profileLength <= 0 || profileLength > MAX_FIELD_LENGTH) {
+        throw new IllegalArgumentException(String.format("Invalid profile length; max allowed '%d', got '%d'", MAX_FIELD_LENGTH, profileLength));
+      }
+
+      // decode the profile name
+      byte[] profileBytes = new byte[profileLength];
+      buffer.get(profileBytes);
+      String profile = new String(profileBytes);
+
+      // validate the entity length
+      int entityLength = buffer.getInt();
+      if (entityLength <= 0 || entityLength > MAX_FIELD_LENGTH) {
+        throw new IllegalArgumentException(String.format("Invalid entity length; max allowed '%d', got '%d'", MAX_FIELD_LENGTH, entityLength));
+      }
+
+      // decode the entity
+      byte[] entityBytes = new byte[entityLength];
+      buffer.get(entityBytes);
+      String entity = new String(entityBytes);
+
+      // decode the groups
+      List<Object> groups = new ArrayList<>();
+      int numberOfGroups = buffer.getInt();
+      for (int i = 0; i < numberOfGroups; i++) {
+
+        // validate the group length
+        int groupLength = buffer.getInt();
+        if (groupLength <= 0 || groupLength > MAX_FIELD_LENGTH) {
+          throw new IllegalArgumentException(String.format("Invalid group length; max allowed '%d', got '%d'", MAX_FIELD_LENGTH, groupLength));
+        }
+
+        // decode the group
+        byte[] groupBytes = new byte[groupLength];
+        buffer.get(groupBytes);
+
+        String group = new String(groupBytes);
+        groups.add(group);
+      }
```

Comment:
An example in support: suppose you decode rowkeys, and see a bunch of entity values that look like "192.168.52.122", "10.5.123.52", "192.147.130.204", "172.217.5.110".  Of course those are probably IP addresses.  But are they source addresses or destination addresses?  There's no way to tell without access to the Profile spec.

Group values, of course, are potentially even worse.